File: 2022/Day7.lua
function getSize(file)
  if file.t == "file" then return file.size end

  local size = 0

  for _,v in pairs(file.children) do
    size = size + getSize(v)
  end

  return size
end

function getAllSizes(current, sizes)
  sizes = sizes or {}

  if current.t == "dir" then
    table.insert(sizes, getSize(current))
  end

  for _,v in pairs(current.children) do
    getAllSizes(v, sizes)
  end
end

local SMALLER_THAN = 100000
local TOTAL_FILESYSTEM_SIZE = 70000000
local TOTAL_UPDATE_SIZE = 30000000

local root = {
  name = "/",
  children = {},
  parent = nil,
  t = "dir",
  size = 0,
}

local current = nil

for line in io.lines() do
  if line:find("$ cd") then
    local name = line:sub(6)

    if name == ".." then
      current = current.parent
    elseif name == "/" then
      current = root
    else
      current = current.children[name]
    end
  elseif not line:find("%$") then
    local first, second = line:match("(.+) (.+)")

    if first:find("dir") then
      if not current.children[second] then
        current.children[second] = {
          name = second,
          children = {},
          parent = current,
          t = "dir",
          size = 0
        }
      end
    else
      current.children[second] = {
        name = second,
        children = {},
        parent = current,
        t = "file",
        size = tonumber(first)
      }
    end
  end
end

local sizes = {}
getAllSizes(root, sizes)

local total = 0

for _,v in pairs(sizes) do
  if v <= SMALLER_THAN then
    total = total + v
  end
end

print(total)

local remainingToUpdate = TOTAL_UPDATE_SIZE - (TOTAL_FILESYSTEM_SIZE - getSize(root))

local min = math.huge

for _,v in pairs(sizes) do
  if v >= remainingToUpdate and v < min then
    min = v
  end
end

print(min)

File: 2022/Day7.cs
using System.Linq;
using System.Collections.Generic;
using System.Threading;

namespace aoc.Year2022;

public class Day7 : IDay {
  private class Dir {
    public string Name { get; set; } = "";
    public Dir? Parent { get; set; }
    public Dictionary<string, Dir> Children = new Dictionary<string, Dir>();
    public Dictionary<string, int> Files = new Dictionary<string, int>();

    public int GetFilesSize() => Files.Aggregate(0, (acc, x) => acc + x.Value);
    public int GetSize() {
      var size = GetFilesSize();

      foreach (var child in Children) {
        size += child.Value.GetSize();
      }

      return size;
    }
  }

  private const int TOTAL_FILESYSTEM_SIZE = 70000000;
  private const int TOTAL_UPDATE_SIZE = 30000000;

  public void PartOne(string input) {
    var root = GetFilesystem(input);

    var sizes = new List<int>();

    GetAllDirectorySizes(sizes, root);

    Console.WriteLine(sizes.Where(x => x <= 100000).Sum());
  }

  public void PartTwo(string input) {
    var root = GetFilesystem(input);

    var rootSize = root.GetSize();

    var diff = TOTAL_FILESYSTEM_SIZE - rootSize;

    var remainingToUpdate = TOTAL_UPDATE_SIZE - diff;

    var sizes = new List<int>();

    GetAllDirectorySizes(sizes, root);

    Console.WriteLine(sizes.Where(x => x >= remainingToUpdate).OrderBy(x => x).First());
  }

  private static Dir GetFilesystem(string input) {
    var root = new Dir {
      Name = "/",
      Parent = null
    };

    root.Parent = root;

    var current = root;

    foreach (var l in input.ToLines()[1..]) {
      if (l.StartsWith("$ cd")) {
        var name = l[5..];

        if (name == ".." && current.Parent != null) current = current.Parent;
        else if (name == "/") current = root;
        else current = current.Children[name];
      } else if (!l.StartsWith("$")) {
        if (current == null) throw new Exception("Huh?");

        var listing = l.Split(" ");

        if (listing[0] == "dir") {
          if (!current.Children.ContainsKey(listing[1])) {
            current.Children[listing[1]] = new Dir {
              Name = listing[1],
              Parent = current,
            };
          }
        } else {
          var size = int.Parse(listing[0]);
          var name = listing[1];

          current.Files[name] = size;
        }
      }
    }

    return root;
  }

  private static void GetAllDirectorySizes(List<int> sizes, Dir current) {
    sizes.Add(current.GetSize());

    foreach (var child in current.Children) {
      GetAllDirectorySizes(sizes, child.Value);
    }
  }
}
